src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs

```diff
@@ -172,6 +172,16 @@ private List<SourceRepository> GetEffectiveSourcesCore(ISettings settings, IList
                 // Create a shared caching provider if one does not exist already
                 CachingSourceProvider = new CachingSourceProvider(packageSourceProvider);
             }
+            else
+            {
+                foreach(var source in CachingSourceProvider.GetRepositories())
```

Comment:
When does this happen? 

I think the caching provider is initialized in the regular restore case isn't it?

Reply:
this happens while creating restoreContext from DependencyGraphRetoreUtility [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L287) 

Then we add the additional source being passed through restore [here](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L290) This additional source only matters for IVS/ template wizard APIs in all other cases, this source is already coming from NuGet.config chain where as we just take the full list of sources because it does a restore (not installing a specific package).

Reply:
There really only is 1 correct way to provider sources to restore as implemented currently and that is through the package spec. 

You basically allow the GetPackageSpecs to figure out the sources. 

There is also another way to do it, through the RestoreArgs - This was kept for backwards compatibility when implemented and it's slowly been removed. Currently it's only productive usage is the Add package dotnet command.
If you look where the sources are set in RestoreArgs it's only tests + AddPackageReferenceCommand.

This here adds a new way to calculate the effective sources that works in an non-obvious way. 
I think there are some unforeseen consequences here. 

The CachingSourceProvider is shared per restore operation as you said. By adding this you are potentially overriding the configs configured by the user through MSBuild properties. You would always have the dg spec sources(which could be msbuild) *and* the config sources that they tried to override. 

My fear is that this changes makes https://github.com/NuGet/NuGet.Client/pull/1976 obsolete, as now the specified source will be respected in addition to the already defined project sources.

Reply:
Given the current scheme of things, I disagree that only correct way to provide sources is through GetPackageSpec because that doesn't cover command line arguments, explicit PMC arguments, or IVS parameters. These all are extension points which should be respected in similar way and shouldn't be discriminated.

> The CachingSourceProvider is shared per restore operation as you said. By adding this you are potentially overriding the configs configured by the user through MSBuild properties. You would always have the dg spec sources(which could be msbuild) and the config sources that they tried to override.

With this PR, we aren't overriding the configs, instead adding the additional source which isn't part of config, there is different between the both. 

For #1976, then I'm not sure if adding that warning was the right thing for #5763. But then we haven't really solved that issue, we just added a warning to tell consumers that it doesn't work.

I still don't see a reasoning of not doing this, may be I'm missing a bigger picture here. IMHO customers are making a conscious choice to install a package from a specific source which isn't part of their configs, knowing the consequences and we should allow it like we allow for `packages.config`.  And there are legit customers reasons for doing that.

Reply:
> Given the current scheme of things, I disagree that only correct way to provide sources is through GetPackageSpec because that doesn't cover command line arguments, explicit PMC arguments, or IVS parameters. These all are extension points which should be respected in similar way and shouldn't be discriminated.

This is exactly how it's implemented though. 
We've time and time again wanted people to avoid using specific sources for PR restores because it breaks repeatability. 

> With this PR, we aren't overriding the configs, instead adding the additional source which isn't part of config, there is different between the both.

There are multiple ways the sources list is populated. 
https://github.com/NuGet/NuGet.Client/blob/736655ce040097eb60e574af72a9d5d0beb248ca/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L280

The sources here are coming from https://github.com/NuGet/NuGet.Client/blob/736655ce040097eb60e574af72a9d5d0beb248ca/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L37, which has lots of calls. 
In most cases this is populated from the setting sources. 

You could've probably avoided the changes in the RestoreArgs.GetEffectiveSources by populating the Sources of the restore args. 

But the same point would still stand in that case. 
It would have unintended consequences where you'd have extra sources in the scenarios where someone has used the RestoreSources msbuild property.

> For #1976, then I'm not sure if adding that warning was the right thing for #5763. But then we haven't really solved that issue, we just added a warning to tell consumers that it doesn't work.

> I still don't see a reasoning of not doing this, may be I'm missing a bigger picture here. IMHO customers are making a conscious choice to install a package from a specific source which isn't part of their configs, knowing the consequences and we should allow it like we allow for packages.config. And there are legit customers reasons for doing that.

The reasoning was that we should not solve it as a "bad practices" thing. (At least for PMC) and the Package Manager UI.

Reply:
> You could've probably avoided the changes in the RestoreArgs.GetEffectiveSources by populating the Sources of the restore args.

This will not work, since `RestoreArgs` sources are only consumed when they are also available in dgSpec source (which means there are also part of packageSpec sources) which is why I didn't add them there. Corresponding source - https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs#L154

Reply:
Read the below code;

Reply:
dah... this comment `//DGSpecSources should always match the Sources` confused me. :)

Reply:
Yours truly is to blame for that 😉